src/Microsoft.AspNet.Routing/RouteValueDictionary.cs

```diff
@@ -289,7 +302,126 @@ public bool TryGetValue(string key, out object value)
                 throw new ArgumentNullException(nameof(key));
             }
 
-            return _dictionary.TryGetValue(key, out value);
+            if (_dictionary != null)
+            {
+                return _dictionary.TryGetValue(key, out value);
+            }
+            else if (_properties != null)
+            {
+                for (var i = 0; i < _properties.Length; i++)
+                {
+                    var property = _properties[i];
+                    if (Comparer.Equals(property.Name, key))
+                    {
+                        value = property.ValueGetter(_value);
+                        return true;
+                    }
+                }
+
+                value = null;
+                return false;
+            }
+            else
+            {
+                value = null;
+                return false;
+            }
+        }
+
+        private void EnsureWritable()
+        {
+            if (_dictionary == null && _properties == null)
+            {
+                _dictionary = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+            }
+            else if (_dictionary == null)
+            {
+                _dictionary = new Dictionary<string, object>(_properties.Length + 1, StringComparer.OrdinalIgnoreCase);
+
+                for (var i = 0; i < _properties.Length; i++)
+                {
+                    var property = _properties[i];
+                    _dictionary.Add(property.Property.Name, property.ValueGetter(_value));
+                }
+            }
+        }
+
+        public struct Enumerator : IEnumerator<KeyValuePair<string, object>>
+        {
+            private readonly RouteValueDictionary _dictionary;
+
+            private int _index;
+            private Dictionary<string, object>.Enumerator _enumerator;
+
+            public Enumerator(RouteValueDictionary dictionary)
+            {
+                if (dictionary == null)
+                {
+                    throw new InvalidOperationException();
+                }
+
+                _dictionary = dictionary;
+
+                Current = default(KeyValuePair<string, object>);
+                _index = -1;
+                _enumerator = _dictionary._dictionary == null ? 
+                    default(Dictionary<string, object>.Enumerator) : 
+                    _dictionary._dictionary.GetEnumerator();
+            }
+
+            public KeyValuePair<string, object> Current { get; private set; }
+
+            object IEnumerator.Current => Current;
+
+            public void Dispose()
+            {
+            }
+
+            public bool MoveNext()
+            {
+                if (_dictionary?._dictionary != null)
+                {
+                    if (_enumerator.MoveNext())
+                    {
+                        Current = _enumerator.Current;
+                        return true;
+                    }
+                    else
```

Comment:
Could this be rewritten without the `else`s here and making the block inside the default fall through case?

````if (_diction?._dictionary != null)
{

}
else if (_dictionary?._properties != null)
{

}

Current = default(KeyValuePair<string, object>);
return false;
```